

## Introduction to Data Exploratory Analysis

Welcome to the Data Exploratory Analysis notebook for the Employee Performance Analysis project at INX Future Inc. This notebook serves as a critical component of our comprehensive analysis, aiming to uncover insights into the factors that influence employee performance.

### Objectives:
- **Understand the Data**: Gain a thorough understanding of the various features within our dataset, such as age, department, job role, and performance ratings, and their distributions.
- **Identify Patterns and Anomalies**: Explore patterns, trends, and potential anomalies in the data that could influence the subsequent phases of modeling and analysis.
- **Prepare for Further Analysis**: Use the insights gained from this exploratory analysis to guide our data preprocessing choices and feature engineering efforts, ensuring that our predictive models are built on a solid foundation of well-understood data.

In this notebook, we will employ a variety of statistical tests and visualization techniques to analyze the dataset. The findings from this exploratory phase will directly inform our approach to building predictive models and making strategic recommendations.

Let’s dive into the data to discover the story it tells about employee performance at INX Future Inc.



In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [2]:
# Load the dataset
file_path = r'C:\Users\moses_y\OneDrive\Desktop\ML Projects\Employee_Performance_Analysis\data\raw\INX_Future_Inc_Employee_Performance_CDS_Project2_Data_V1.8.xlsx'

xls = pd.ExcelFile(file_path)

# Load each sheet into a DataFrame
df1 = pd.read_excel(xls, sheet_name=0)
df2 = pd.read_excel(xls, sheet_name=1)


In [3]:

# Display the first few rows of each sheet to understand their structure and content
df1_head = df1.head()
df1_head

,EmpNumber,Age,Gender,EducationBackground,MaritalStatus,EmpDepartment,EmpJobRole,BusinessTravelFrequency,DistanceFromHome,EmpEducationLevel,...,EmpRelationshipSatisfaction,TotalWorkExperienceInYears,TrainingTimesLastYear,EmpWorkLifeBalance,ExperienceYearsAtThisCompany,ExperienceYearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager,Attrition,PerformanceRating
0,E1001000,32,Male,Marketing,Single,Sales,Sales Executive,Travel_Rarely,10,3,...,4,10,2,2,10,7,0,8,No,3
1,E1001006,47,Male,Marketing,Single,Sales,Sales Executive,Travel_Rarely,14,4,...,4,20,2,3,7,7,1,7,No,3
2,E1001007,40,Male,Life Sciences,Married,Sales,Sales Executive,Travel_Frequently,5,4,...,3,20,2,3,18,13,1,12,No,4
3,E1001009,41,Male,Human Resources,Divorced,Human Resources,Manager,Travel_Rarely,10,4,...,2,23,2,2,21,6,12,6,No,3
4,E1001010,60,Male,Marketing,Single,Sales,Sales Executive,Travel_Rarely,16,4,...,4,10,1,3,2,2,2,2,No,3


The data has been successfully loaded, and here's an overview of the first few rows:

| EmpNumber | Age | Gender | EducationBackground | MaritalStatus | EmpDepartment | EmpJobRole | BusinessTravelFrequency | DistanceFromHome | EmpEducationLevel | ... | PerformanceRating |
|-----------|-----|--------|---------------------|---------------|---------------|------------|-------------------------|------------------|-------------------|-----|-------------------|
| E1001000  | 32  | Male   | Marketing           | Single        | Sales         | Sales Executive | Travel_Rarely          | 10               | 3                 | ... | 3                 |
| E1001006  | 47  | Male   | Marketing           | Single        | Sales         | Sales Executive | Travel_Rarely          | 14               | 4                 | ... | 3                 |
| E1001007  | 40  | Male   | Life Sciences       | Married       | Sales         | Sales Executive | Travel_Frequently      | 5                | 4                 | ... | 4                 |
| E1001009  | 41  | Male   | Human Resources     | Divorced      | Human Resources | Manager         | Travel_Rarely          | 10               | 4                 | ... | 3                 |
| E1001010  | 60  | Male   | Marketing           | Single        | Sales         | Sales Executive | Travel_Rarely          | 16               | 4                 | ... | 3                 |

This dataset includes various attributes of employees such as age, gender, education background, marital status, department, job role, business travel frequency, distance from home, education level, and performance rating, among others. 



In [4]:
df2_head = df2.head()
df2_head

,Unnamed: 0,Unnamed: 1
0,NaN,NaN
1,EmpEducationLevel,1 'Below College'
2,NaN,2 'College'
3,NaN,3 'Bachelor'
4,NaN,4 'Master'


Printing out all columns before starting with visualizations is a good practice, especially in comprehensive Exploratory Data Analysis (EDA). It gives you a clearer overview of the dataset, helping you understand the type of data each column holds, which is crucial for making informed decisions on which visualizations will be most effective. Knowing all the columns can help you identify:

- **Categorical Variables:** For which you might use count plots, bar plots, or pie charts to understand their distribution.
- **Numerical Variables:** Where histograms, density plots, or scatter plots can reveal distributions and relationships.
- **Text Variables:** Which might require specific NLP techniques if they are to be analyzed or used in modeling.
- **Date/Time Variables:** Time series analysis might be warranted for these, with line plots to show trends over time.

Let's print out all the columns from the dataset to have a clear view of what we're working with.



In [16]:
# Printing all columns of the raw dataset to understand its structure better
raw_data_columns = df1.columns.tolist()
raw_data_columns


['EmpNumber',
 'Age',
 'Gender',
 'EducationBackground',
 'MaritalStatus',
 'EmpDepartment',
 'EmpJobRole',
 'BusinessTravelFrequency',
 'DistanceFromHome',
 'EmpEducationLevel',
 'EmpEnvironmentSatisfaction',
 'EmpHourlyRate',
 'EmpJobInvolvement',
 'EmpJobLevel',
 'EmpJobSatisfaction',
 'NumCompaniesWorked',
 'OverTime',
 'EmpLastSalaryHikePercent',
 'EmpRelationshipSatisfaction',
 'TotalWorkExperienceInYears',
 'TrainingTimesLastYear',
 'EmpWorkLifeBalance',
 'ExperienceYearsAtThisCompany',
 'ExperienceYearsInCurrentRole',
 'YearsSinceLastPromotion',
 'YearsWithCurrManager',
 'Attrition',
 'PerformanceRating']

Here are all the columns in the raw dataset:

1. **EmpNumber:** Employee ID or number.
2. **Age:** Age of the employee.
3. **Gender:** Gender of the employee.
4. **EducationBackground:** Educational background of the employee.
5. **MaritalStatus:** Marital status of the employee.
6. **EmpDepartment:** Department of the employee.
7. **EmpJobRole:** Job role of the employee.
8. **BusinessTravelFrequency:** Frequency of business travel.
9. **DistanceFromHome:** Distance from home to work.
10. **EmpEducationLevel:** Education level of the employee.
11. **EmpEnvironmentSatisfaction:** Employee's satisfaction with the work environment.
12. **EmpHourlyRate:** Hourly rate of the employee.
13. **EmpJobInvolvement:** Employee's involvement in their job.
14. **EmpJobLevel:** Job level of the employee.
15. **EmpJobSatisfaction:** Employee's job satisfaction.
16. **NumCompaniesWorked:** Number of companies the employee has worked at.
17. **OverTime:** Whether the employee works overtime.
18. **EmpLastSalaryHikePercent:** The percentage of the last salary hike.
19. **EmpRelationshipSatisfaction:** Employee's satisfaction with their relationships at work.
20. **TotalWorkExperienceInYears:** Total work experience in years.
21. **TrainingTimesLastYear:** Number of trainings last year.
22. **EmpWorkLifeBalance:** Employee's work-life balance.
23. **ExperienceYearsAtThisCompany:** Number of years the employee has worked at this company.
24. **ExperienceYearsInCurrentRole:** Number of years in the current role.
25. **YearsSinceLastPromotion:** Number of years since the last promotion.
26. **YearsWithCurrManager:** Number of years with the current manager.
27. **Attrition:** Whether the employee has left the company or not.
28. **PerformanceRating:** The performance rating of the employee.

Having this list enables us to make more targeted decisions regarding which variables might be of interest for deeper analysis and visualization. For example, examining the relationship between `PerformanceRating` and variables like `EmpDepartment`, `Age`, `TotalWorkExperienceInYears`, or `EmpJobSatisfaction` could yield insights into factors influencing employee performance. Additionally, understanding the distribution of `EducationBackground`, `EmpDepartment`, or `Gender` across the company could help in diversity analysis.

In [5]:
# Check for missing values in both DataFrames
missing_values_df1 = df1.isnull().sum()

# duplicates
duplicates_df1 = df1.duplicated().sum()



(missing_values_df1,missing_values_df1)


(EmpNumber                       0
 Age                             0
 Gender                          0
 EducationBackground             0
 MaritalStatus                   0
 EmpDepartment                   0
 EmpJobRole                      0
 BusinessTravelFrequency         0
 DistanceFromHome                0
 EmpEducationLevel               0
 EmpEnvironmentSatisfaction      0
 EmpHourlyRate                   0
 EmpJobInvolvement               0
 EmpJobLevel                     0
 EmpJobSatisfaction              0
 NumCompaniesWorked              0
 OverTime                        0
 EmpLastSalaryHikePercent        0
 EmpRelationshipSatisfaction     0
 TotalWorkExperienceInYears      0
 TrainingTimesLastYear           0
 EmpWorkLifeBalance              0
 ExperienceYearsAtThisCompany    0
 ExperienceYearsInCurrentRole    0
 YearsSinceLastPromotion         0
 YearsWithCurrManager            0
 Attrition                       0
 PerformanceRating               0
 dtype: int64,
 EmpN

### Missing Values:
- The dataset does not contain any missing values across all its features, indicating that it is clean and ready for further analysis without the need for imputation.

In [13]:
# Summary statistics for numerical features
numerical_summary = df1.describe()

# Defining categorical columns 
categorical_cols = df1.select_dtypes(include=['object']).columns

# Distribution of categorical variables
categorical_distribution = {column: df1[column].value_counts() for column in categorical_cols}

# Definition of numerical_cols
numerical_cols = df1.select_dtypes(include=['int64', 'float64']).columns
# Correlation matrix for numerical features
corr_matrix = df1[numerical_cols].corr()


numerical_summary

,Age,DistanceFromHome,EmpEducationLevel,EmpEnvironmentSatisfaction,EmpHourlyRate,EmpJobInvolvement,EmpJobLevel,EmpJobSatisfaction,NumCompaniesWorked,EmpLastSalaryHikePercent,EmpRelationshipSatisfaction,TotalWorkExperienceInYears,TrainingTimesLastYear,EmpWorkLifeBalance,ExperienceYearsAtThisCompany,ExperienceYearsInCurrentRole,YearsSinceLastPromotion,YearsWithCurrManager,PerformanceRating
count,1200.000000,1200.000000,1200.00000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000,1200.000000
mean,36.918333,9.165833,2.89250,2.715833,65.981667,2.731667,2.067500,2.732500,2.665000,15.222500,2.725000,11.330000,2.785833,2.744167,7.077500,4.291667,2.194167,4.105000,2.948333
std,9.087289,8.176636,1.04412,1.090599,20.211302,0.707164,1.107836,1.100888,2.469384,3.625918,1.075642,7.797228,1.263446,0.699374,6.236899,3.613744,3.221560,3.541576,0.518866
min,18.000000,1.000000,1.00000,1.000000,30.000000,1.000000,1.000000,1.000000,0.000000,11.000000,1.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,2.000000
25%,30.000000,2.000000,2.00000,2.000000,48.000000,2.000000,1.000000,2.000000,1.000000,12.000000,2.000000,6.000000,2.000000,2.000000,3.000000,2.000000,0.000000,2.000000,3.000000
50%,36.000000,7.000000,3.00000,3.000000,66.000000,3.000000,2.000000,3.000000,2.000000,14.000000,3.000000,10.000000,3.000000,3.000000,5.000000,3.000000,1.000000,3.000000,3.000000
75%,43.000000,14.000000,4.00000,4.000000,83.000000,3.000000,3.000000,4.000000,4.000000,18.000000,4.000000,15.000000,3.000000,3.000000,10.000000,7.000000,3.000000,7.000000,3.000000
max,60.000000,29.000000,5.00000,4.000000,100.000000,4.000000,5.000000,4.000000,9.000000,25.000000,4.000000,40.000000,6.000000,4.000000,40.000000,18.000000,15.000000,17.000000,4.000000



### Summary Statistics for Numerical Features:
- **Age:** Employees range from 18 to 60 years old, with an average age of approximately 37 years.
- **DistanceFromHome, EmpEducationLevel, EmpEnvironmentSatisfaction,** and other fields provide a comprehensive view of the workforce's demographic, educational, and job satisfaction levels.

In [14]:
categorical_distribution

{'EmpNumber': EmpNumber
 E1001000    1
 E100346     1
 E100342     1
 E100341     1
 E100340     1
            ..
 E1001718    1
 E1001717    1
 E1001716    1
 E1001713    1
 E100998     1
 Name: count, Length: 1200, dtype: int64,
 'Gender': Gender
 Male      725
 Female    475
 Name: count, dtype: int64,
 'EducationBackground': EducationBackground
 Life Sciences       492
 Medical             384
 Marketing           137
 Technical Degree    100
 Other                66
 Human Resources      21
 Name: count, dtype: int64,
 'MaritalStatus': MaritalStatus
 Married     548
 Single      384
 Divorced    268
 Name: count, dtype: int64,
 'EmpDepartment': EmpDepartment
 Sales                     373
 Development               361
 Research & Development    343
 Human Resources            54
 Finance                    49
 Data Science               20
 Name: count, dtype: int64,
 'EmpJobRole': EmpJobRole
 Sales Executive              270
 Developer                    236
 Manager R&D        

### Distribution of Categorical Variables:
- **Gender:** There are more males (725) than females (475) in the dataset.
- **EmpDepartment:** The Sales, Development, and Research & Development departments are the largest, with Sales leading at 373 employees.
- **EducationBackground:** A significant number of employees come from Life Sciences (492) and Medical (384) backgrounds.
- **MaritalStatus:** A majority of the employees are Married (548), followed by Single (384) and Divorced (268).
- **BusinessTravelFrequency:** Most employees rarely travel (846), with a smaller number frequently traveling (222) or not traveling at all (132).
- **OverTime:** A larger portion of employees do not work overtime (847) compared to those who do (353).
- **Attrition:** The majority of employees have not experienced attrition (1022), while a smaller number have (178).

## Deeper Statistical Analysis
This step will focus on understanding the statistical significance of various factors affecting employee performance. Techniques such as ANOVA (for categorical variables) and regression analysis (for continuous variables) can be employed to assess the impact of different variables on the performance rating.

Implementation Plan:

- ANOVA Test: To compare means across different groups (e.g., departments, education background) to see if there is a statistically significant difference in performance ratings.

In [17]:
import scipy.stats as stats

# Preparing data for ANOVA test: Performance Rating by Department
anova_data = df1[['EmpDepartment', 'PerformanceRating']].copy()
groups = [group["PerformanceRating"].values for _, group in anova_data.groupby("EmpDepartment")]

# Perform ANOVA
f_value, p_value = stats.f_oneway(*groups)
f_value, p_value


(8.940464532824977, 2.3812439958733325e-08)

The ANOVA test on Performance Rating by Department yielded an F-Value of approximately 8.94 and a P-Value of roughly \(2.38 \times 10^{-8}\).

### Interpretation:
- **F-Value:** The obtained F-Value suggests that there are some differences among the means of the groups. Higher F-Values generally indicate a stronger degree of variance among the group means.
- **P-Value:** The P-Value is significantly less than 0.05 (common threshold for significance), indicating that the differences in performance ratings across departments are statistically significant.

### What This Means:
The statistical analysis reveals that there are indeed significant differences in employee performance ratings across different departments within INX Future Inc. This finding suggests that departmental factors may play a crucial role in influencing employee performance.



## Regression Analysis
Let's start by quantifying the relationship between employee performance (PerformanceRating) and various predictors. For simplicity and demonstrability, we'll use a linear regression model, even though the performance rating is categorical, to illustrate the process. In practice, ordinal regression or classification models might be more appropriate.

In [18]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import LabelEncoder

# Preparing data for Regression Analysis
# Encoding categorical variables for simplicity
le = LabelEncoder()
encoded_data = df1.apply(le.fit_transform)

# Selecting predictors - for the illustration we'll limit the number of variables
predictors = ['Age', 'DistanceFromHome', 'EmpEducationLevel', 'EmpEnvironmentSatisfaction', 
              'EmpJobInvolvement', 'EmpJobLevel', 'EmpJobSatisfaction', 'NumCompaniesWorked', 
              'EmpLastSalaryHikePercent', 'EmpWorkLifeBalance', 'TotalWorkExperienceInYears', 
              'TrainingTimesLastYear', 'YearsSinceLastPromotion', 'YearsWithCurrManager']
target = 'PerformanceRating'

X = encoded_data[predictors]
y = encoded_data[target]

# Splitting data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Linear Regression Model
model = LinearRegression()
model.fit(X_train, y_train)

# Predicting on test data
y_pred = model.predict(X_test)

# Evaluating the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

mse, r2


(0.17255801215915695, 0.2602454971444298)

## Regression Analysis Results:
Mean Squared Error (MSE): The model has a MSE of approximately 0.173, indicating the average squared difference between the estimated values and the actual value.
R-squared (R²): The R² score is approximately 0.260, which represents the proportion of the variance for the dependent variable that's explained by the independent variables in the model.

## Interpretation:
The R² value suggests that the model explains around 26% of the variability in the performance ratings, which is relatively low. This indicates that while the selected predictors do have some relationship with performance rating, there are likely other factors not captured by this model that influence employee performance.
The MSE value provides a measure of the model's accuracy in predicting performance ratings. 

`A lower MSE would be preferred, indicating better model performance.`

The Exploratory Data Analysis (EDA) of the raw data from INX Future Inc. Employee Performance dataset has provided several insights:



### Insights and Next Steps:

- The distribution and correlation analyses suggest specific areas for deeper dives, such as understanding the impact of department, age, gender, or work experience on employee performance.
- With the comprehensive EDA complete, focusing on feature selection and engineering based on these insights can better prepare the dataset for modeling. The correlations and distributions observed should guide the development of predictive models aimed at analyzing employee performance.

This EDA provides a solid foundation for moving forward with the project, enabling informed decisions for feature engineering, model selection, and ultimately, insights into factors influencing employee performance.

In [20]:
# Detailed Department Analysis

raw_data = df1
# Analyzing the mean performance rating by department
department_performance = raw_data.groupby('EmpDepartment')['PerformanceRating'].mean().sort_values()

# Employee satisfaction levels by department
environment_satisfaction = raw_data.groupby('EmpDepartment')['EmpEnvironmentSatisfaction'].mean()
job_satisfaction = raw_data.groupby('EmpDepartment')['EmpJobSatisfaction'].mean()
work_life_balance = raw_data.groupby('EmpDepartment')['EmpWorkLifeBalance'].mean()

# Combining the satisfaction levels for a comprehensive view
satisfaction_levels = pd.DataFrame({
    'EnvironmentSatisfaction': environment_satisfaction, 
    'JobSatisfaction': job_satisfaction, 
    'WorkLifeBalance': work_life_balance
})

department_performance


EmpDepartment
Finance                   2.775510
Sales                     2.860590
Research & Development    2.921283
Human Resources           2.925926
Data Science              3.050000
Development               3.085873
Name: PerformanceRating, dtype: float64

In [21]:
satisfaction_levels

,EnvironmentSatisfaction,JobSatisfaction,WorkLifeBalance
EmpDepartment,,,
Data Science,2.950000,2.700000,2.500000
Development,2.709141,2.747922,2.686981
Finance,2.836735,2.653061,2.673469
Human Resources,2.629630,2.555556,2.925926
Research & Development,2.775510,2.720117,2.743440
Sales,2.651475,2.766756,2.796247


### Detailed Department Analysis Results:

#### Average Performance Rating by Department:
- The analysis reveals variance in average performance ratings across departments, with **Data Science** and **Development** departments showing the highest average performance ratings, whereas **Finance** and **Sales** have lower averages. This suggests that certain departmental factors or practices may significantly impact employee performance.

#### Employee Satisfaction Levels by Department:
- **Environment Satisfaction, Job Satisfaction, and Work-Life Balance** vary across departments:
  - **Data Science** scores high in environment satisfaction but has lower job satisfaction compared to other departments.
  - **Development** shows relatively balanced satisfaction levels across the three categories.
  - **Human Resources** has the highest work-life balance satisfaction, despite its middle-range performance rating.
  - **Sales**, despite its lower performance rating average, has relatively high job satisfaction.

### Interpretation:
- The variance in performance ratings and satisfaction levels across departments underscores the importance of department-specific practices and policies on employee performance and satisfaction.
- High satisfaction levels in certain departments do not necessarily correlate with high performance ratings, suggesting that other factors (such as managerial practices, department goals, or individual employee roles) significantly influence performance.

### Strategic Implications:
- **Tailored Improvement Plans:** Each department may require tailored strategies that address specific areas of improvement. For instance, initiatives to boost job satisfaction in Data Science could enhance overall performance, while in Sales, focusing on environment satisfaction might yield better outcomes.
- **Focus on Work-Life Balance:** Departments like Human Resources, which show high work-life balance satisfaction, can serve as benchmarks for best practices that could be implemented across other departments to improve overall satisfaction and performance.
- **Cross-Departmental Learning:** Departments with high performance ratings (such as Data Science and Development) might share effective practices and policies with lower-performing departments.

This detailed departmental analysis provides critical insights into the drivers of employee performance and satisfaction at INX Future Inc., laying a solid foundation for developing targeted strategies to enhance overall performance.

In [19]:
# Performing correlation analysis to understand the relationship between features and employee performance
correlation_matrix = df1.corr()

# Focusing on the correlation with PerformanceRating
performance_correlation = correlation_matrix['PerformanceRating'].sort_values()

# Display the correlations sorted by their relationship with PerformanceRating
performance_correlation


ValueError: could not convert string to float: 'E1001000'